### PR TITLE
Fix -incomplete-uni-patterns

### LINKED ISSUE #590: incomplete-uni-patterns
There are a few incomplete-uni-patterns in this library. Since this warning will soon be enabled by -Wall, we should either a) use `-Wno-...` , or b) case+error.

(1)
[Data.Sequence.Internal:L3692](https://github.com/haskell/containers/blob/master/Data/Sequence/Internal.hs#L3692)
```
    Pattern match(es) are non-exhaustive
    In a pattern binding: Patterns not matched: EmptyLTree
     |
3692 |     f' ms = let ConsLTree node m' = viewLTree ms in
     |             
```

(2)
 
[Data.Sequence.Internal:L3707](https://github.com/haskell/containers/blob/master/Data/Sequence/Internal.hs#L3707)

```
    Pattern match(es) are non-exhaustive
    In a pattern binding: Patterns not matched: EmptyRTree
     |
3707 |     f' ms =  let SnocRTree m' node = viewRTree ms in
     |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

(3) 

[Data.IntMap.Internal:L1049](https://github.com/haskell/containers/blob/master/Data/IntMap/Internal.hs#L1049)
```
    Pattern match(es) are non-exhaustive
    In a lambda abstraction:
        Patterns not matched:
            (Bin _ _ _ _) _
            (Tip _ _) (Bin _ _ _ _)
            (Tip _ _) Nil
            Nil _
     |
1049 |   = mergeWithKey' Bin (\(Tip k1 x1) (Tip _k2 x2) -> Tip k1 (f k1 x1 x2)) id id m1 m2
     |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

(4)

[Data.IntMap.Internal:L1246](https://github.com/haskell/containers/blob/master/Data/IntMap/Internal.hs#L1246)

```
    Pattern match(es) are non-exhaustive
    In a lambda abstraction:
        Patterns not matched:
            (Bin _ _ _ _) _
            (Tip _ _) (Bin _ _ _ _)
            (Tip _ _) Nil
            Nil _
    |
715 |   = mergeWithKey' bin (\(Tip k1 x1) (Tip _k2 x2) -> Tip k1 $! f k1 x1 x2) (const Nil) (const Nil) m1 m2
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


```

(5)
[Data.IntMap.Internal:L1292](https://github.com/haskell/containers/blob/master/Data/IntMap/Internal.hs#L1292)

```
Pattern match(es) are non-exhaustive
    In a lambda abstraction:
        Patterns not matched:
            (Bin _ _ _ _) _
            (Tip _ _) (Bin _ _ _ _)
            (Tip _ _) Nil
            Nil _
     |
1292 |         combine = \(Tip k1 x1) (Tip _k2 x2) ->
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

### PR DESCRIPTION
**Pull request by ckoparkar**
#590.

### CODE REVIEW

File: Data/IntMap/Strict.hs [hunk 1]
```diff
@@ -758,8 +764,9 @@ mergeWithKey :: (Key -> a -> b -> Maybe c) -> (IntMap a -> IntMap c) -> (IntMap
              -> IntMap a -> IntMap b -> IntMap c
 mergeWithKey f g1 g2 = mergeWithKey' bin combine g1 g2
   where -- We use the lambda form to avoid non-exhaustive pattern matches warning.
```
Comment (treeowl) [inline Data/IntMap/Strict.hs]: This comment about non-exhaustive patterns should be removed.

**Comment by ckoparkar** [comment]
> Have you tried comparing the merge vs. mergeWithKey' approaches? 

I thought the plan was to apply this simple case+error solution before that. I'll try out the `merge vs. mergeWithKey'` idea and update this patch in the next couple days.

[Changes planned]

**Comment by treeowl** [comment]
I would like to make a release in the next week or so, now that #593 is fixed, but we can certainly wait a couple days.

**Comment by ckoparkar** [comment]
Hi @treeowl, I wanted to get this done this week, but it didn't happen. I spent a little bit of time trying to understand various parts of the `Data.Sequence` implementation. I'll finish it soon, over the weekend perhaps, but please don't let this block a release :)

**Comment by treeowl** [comment]
Oops, I didn't make a release as intended.... Can you rebase this and look into my question?

**Comment by ckoparkar** [comment]
@treeowl sorry for squatting on this for so long. I'll pick this up and try to finish it by this weekend.

**Comment by mitchellwrosen** [comment]
@ckoparkar Ping!

**Comment by sjakobi** [comment]
My impression is that there's very little left to do before this could be merged. Or maybe we could at least merge the non-controversial bits and leave out the "`merge` vs. `mergeWithKey'`" "controversy". (I have no idea what it's about.)

The PR definitely needs a rebase though.

@ckoparkar Can you bring this to completion?

Or would you, @mitchellwrosen, possibly volunteer to finish this?

**Comment by ckoparkar** [comment]
@sjakobi I'm afraid I've forgotten most of the context here. As far as I can tell, David's suggestion was to try to use `merge` instead of `mergeWithKey'` in some of these places. Sorry, but I won't be able to contribute much here. If you or someone else wants to pick this up, please feel free.

**Comment by treeowl** [comment]
`mergeWithKey` is a nasty API that basically forces users to supply partial functions and also allows the data structure invariants to be violated. `merge` is slightly less general API that nevertheless supports pretty much everything one might actually want from `mergeWithKey`. If we can use it without losing performance, that would be a good way to be rid of some incomplete patterns.

**Comment by sjakobi** [comment]
Thanks for the update, @ckoparkar, and thanks for the explanation, @treeowl! :)

I generally agree that it would be nice to use `merge` instead of `mergeWithKey'`. It's far from obvious to me though, that this could be done here without loss of performance: `IntMap.merge` is just `runIdentity (mergeA ...)` and `mergeA` has to do quite a bit of extra work to get the effects sequencing right. `mergeWithKey'` also offers a small optimization specifically for `unionWithKey`:

https://github.com/haskell/containers/blob/4ac960a01f058ae923cf51c6108aae0acbb802dd/containers/src/Data/IntMap/Internal.hs#L1372-L1375

It would of course still be interesting to benchmark results using `merge`, but I think the chances that the results would be good are fairly low.

I also kind of feel that contributors shouldn't be expected to solve a different problem than they originally intended to work on.

So, what I see that _needs_ to be done for this PR to be accepted is:

1. A rebase.
2. Remove the outdated comments in `mergeWithKey` that @treeowl pointed out above.

**Pull request closed without merging**